### PR TITLE
feat: bypass feature flag

### LINKED ISSUE #2704: Add a way to check feature flags with some type-safety
### Preliminary Checks

- [X] I have reviewed https://unkey.com/docs for existing features that would solve my problem

- [X] I have searched for existing feature requests: https://github.com/unkeyed/unkey/issues

- [X] This issue is not a question, general help request, or anything other than a feature request directly related to Unkey. Please ask questions in our Discord community: https://unkey.com/discord.

### Is your feature request related to a problem? Please describe.

Developers need to modify the database for feature flags during local development, which creates unnecessary friction. A type-safe checkFlag function could bypass this requirement in development mode while maintaining proper flag checking in production.

Example:
```ts
  if (!workspace.betaFeatures.logsPage) {
    return notFound();
  }
```

### Describe the solution

We could create a `checkFlag` function that takes two parameters: a feature flag name and a workspace object. This would provide several things:

1. Type safety - by using a type like:
```typescript
type ConfigObject = Extract<typeof workspaces['betaFeatures']['default'], {
   rbac?: boolean;
   ratelimit?: boolean; 
   identities?: boolean;
   logsPage?: boolean;
}>;
```

2. Easy bypassing of feature flags in development:
```typescript
function checkFlag<T extends keyof ConfigObject>(
    flagName: T, 
    workspace
): boolean {
    if (process.env.NODE_ENV === "development") {
        return true
    }
    // Normal flag checking logic...
}
```

This pattern would help ensure we're only checking for flags that actually exist in our configuration, while making it simple to enable all features in development environments.

### Describe alternatives you have considered (if any)

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ogzhanolguncu**
## What does this PR do?

This PR allows us to bypass features and betaFeatures checks when developing locally. It also allows us to pick any feature flag with type safety.

Fixes #2704 (issue)

*If there is not an issue for this, please create one first. This is used to tracking purposes and also helps use understand why this PR exists*

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- Try to navigate to /logs page in local env. Normally this is not possible without manually updating the db.

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [Contributing Guide](./CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand areas
- [x] Ran `pnpm build`
- [x] Ran `pnpm fmt`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [] If a UI change was made: Added a screen recording or screenshots to this PR
- [] Updated the Unkey Docs if changes were necessary

## Release Notes

- **New Features**
	- Enhanced validation and updating mechanisms for the API settings page, ensuring accurate and current data display.
	- Introduced a centralized access control function, `getFlag`, improving clarity and maintainability across various components, including navigation and feature visibility.
	- Added default constants for rate limiting and audit log retention periods to streamline configuration.

- **Bug Fixes**
	- Improved access checks for features like IP whitelisting and rate limit overrides, ensuring users see relevant options based on permissions.

- **Documentation**
	- Updated type definitions and function signatures for better readability and understanding.

### CODE REVIEW

File: apps/dashboard/lib/trpc/routers/ratelimit/createOverride.ts [hunk 1]
```diff
@@ -59,10 +60,7 @@ export const createOverride = t.procedure
             ),
           )
           .then((res) => Number(res.at(0)?.count ?? 0));
-        const max =
-          typeof namespace.workspace.features.ratelimitOverrides === "number"
-            ? namespace.workspace.features.ratelimitOverrides
-            : 5;
+        const max = hasWorkspaceAccess("ratelimitOverrides", namespace.workspace) || 5;
```
[the remark this answers is not in the document]
Comment (ogzhanolguncu) [reply apps/dashboard/lib/trpc/routers/ratelimit/createOverride.ts:63]: Yes, It's the intended behavior

File: apps/dashboard/lib/utils.ts [hunk 2]
```diff
@@ -22,3 +24,59 @@ export function debounce<T extends (...args: any[]) => any>(func: T, delay: numb
...
+ * }
+ *
+ * // Check if workspace has access to a feature with numeric value
+ * const userLimit = hasWorkspaceAccess("userLimit", workspace);
+ * if (userLimit === undefined) {
+ *   return notFound();
+ * }
+ * ```
+ */
+export function hasWorkspaceAccess<T extends keyof ConfigObject>(
```
Comment (chronark) [inline apps/dashboard/lib/utils.ts:57]: nitpick, but the name is a bit weird when returning numbers like retention.

we could rename it to `flag`?
```ts
const days = flag("auditLogRetentionDays", ws)
```

File: apps/dashboard/lib/utils.ts [hunk 3]
```diff
@@ -22,3 +24,59 @@ export function debounce<T extends (...args: any[]) => any>(func: T, delay: numb
...
+  flagName: T,
+  workspace: Partial<WorkspaceFeatures>,
+): FlagValue<T> | null {
+  if (process.env.NODE_ENV === "development") {
```
Comment (chronark) [inline apps/dashboard/lib/utils.ts:61]: how will this work for numbers?

File: apps/dashboard/lib/utils.ts (hunk 2 above)
Comment (ogzhanolguncu) [reply apps/dashboard/lib/utils.ts:57]: Works for me

File: apps/dashboard/lib/utils.ts (hunk 3 above)
Comment (ogzhanolguncu) [reply apps/dashboard/lib/utils.ts:61]: This is just doing a lookup in a type safe manner. If what you are trying to look up is number it casts it to number.

Comment (ogzhanolguncu) [reply apps/dashboard/lib/utils.ts:61]: For reference,

```ts
type ConfigObject = WorkspaceFeatures["betaFeatures"] & WorkspaceFeatures["features"];
// Doing type look-up here
type FlagValue<T extends keyof ConfigObject> = ConfigObject[T];
```

File: apps/dashboard/lib/utils.ts (hunk 2 above)
Comment (ogzhanolguncu) [reply apps/dashboard/lib/utils.ts]: I did, `getFlag` instead since this is a function

File: apps/dashboard/lib/utils.ts (hunk 3 above)
Comment (chronark) [reply apps/dashboard/lib/utils.ts:61]: function flag("retentionDays", { devModeDefault: 30})
function flag("logsPage", { devModeDefault: true})

File: apps/dashboard/app/(app)/identities/layout.tsx [hunk 4]
```diff
@@ -21,7 +22,7 @@ export default async function AuthorizationLayout({
     return redirect("/auth/sign-in");
   }
 
-  if (!workspace.betaFeatures.identities) {
+  if (getFlag(workspace, "identities", { prodFallback: true, devFallback: false })) {
```
Comment (chronark) [inline apps/dashboard/app/(app)/identities/layout.tsx:25]: you flipped it around, so when they had it set to true in our database, we would actually deny them access
```suggestion
  if (!getFlag(workspace, "identities", { prodFallback: false, devFallback: true })) {
```

File: apps/dashboard/app/(app)/logs/page.tsx [hunk 5]
```diff
@@ -28,7 +29,7 @@ export default async function Page({
     return <div>Workspace with tenantId: {tenantId} not found</div>;
   }
 
-  if (!workspace.betaFeatures.logsPage) {
+  if (getFlag(workspace, "logsPage", { devFallback: false, prodFallback: true })) {
```
Comment (chronark) [inline apps/dashboard/app/(app)/logs/page.tsx:32]:
```suggestion
  if (!getFlag(workspace, "logsPage", { devFallback: true, prodFallback: false })) {
```

File: apps/dashboard/app/(app)/workspace-navigations.tsx [hunk 6]
```diff
@@ -91,22 +90,31 @@ export const createWorkspaceNavigation = (
       href: "/monitors/verifications",
       label: "Monitors",
       active: segments.at(0) === "verifications",
-      hidden: !workspace.features.webhooks,
+      hidden: getFlag(workspace, "webhooks", {
```
Comment (chronark) [inline apps/dashboard/app/(app)/workspace-navigations.tsx:93]: that's reversed too
it's too confusing this way I think
in the db I'd set it to `{ "webhooks": true }` if I want to give someone access, but now that would revoke access...

**Review by chronark — CHANGES_REQUESTED** [review]
most of the flags are inverted now, which actually breaks our behavior, I didn't comment on all of them

**Pull request closed without merging**